homeassistant/components/opower/coordinator.py

```diff
@@ -187,59 +192,68 @@ async def _insert_statistics(self) -> None:
                 self.hass, consumption_metadata, consumption_statistics
             )
 
-    async def _async_get_all_cost_reads(self, account: Account) -> list[CostRead]:
-        """Get all cost reads since account activation but at different resolutions depending on age.
+    async def _async_get_cost_reads(
+        self, account: Account, time_zone_str: str, start_time: float | None = None
+    ) -> list[CostRead]:
+        """Get cost reads.
+
+        If start_time is None, get cost reads since account activation,
+        otherwise since start_time - 30 days to allow corrections in data from utilities
 
+        We read at different resolutions depending on age:
         - month resolution for all years (since account activation)
         - day resolution for past 3 years (if account's read resolution supports it)
         - hour resolution for past 2 months (if account's read resolution supports it)
         """
-        cost_reads = []
 
-        start = None
-        end = datetime.now()
-        if account.read_resolution != ReadResolution.BILLING:
-            end -= timedelta(days=3 * 365)
-        cost_reads += await self.api.async_get_cost_reads(
+        def _update_with_finer_cost_reads(
+            cost_reads: list[CostRead], finer_cost_reads: list[CostRead]
+        ) -> None:
+            for i, cost_read in enumerate(cost_reads):
+                for j, finer_cost_read in enumerate(finer_cost_reads):
```

Comment:
It's actually O( len(cost_reads) + len(finer_cost_reads) ) because both lists are sorted and because of the way I'm using break and returns.

The worst will be the very first time when we merge daily with hourly.
We fetch:
daily for the past 3 * 365 days = 1095 items
hourly for the past 2 * 30 * 24 hours = 1440 items

The expectation is to replace the last 60 items of the daily data with all the items of the hourly data.
We iterate over the daily and we quickly skip the first 1095 - 60 items because of the check at line 220 and the break at the next line. This is 1095 - 60 steps
Then either line 215 or 218 will copy all hourly data. This is 1440 steps.
We iterate over the hourly data (the nested for) just in case the results don't start at 00:00. So at worse it will be 23 additional steps.

I could make it a bit faster by iterating the outer loop in reverse order but the code will be less readable so I don't think it's worth it.